AsyncDisplayKit/Private/ASElementMap.m

```diff
@@ -0,0 +1,151 @@
+//
+//  ASElementMap.m
+//  AsyncDisplayKit
+//
+//  Created by Adlai Holler on 2/22/17.
+//  Copyright © 2017 Facebook. All rights reserved.
+//
+
+#import "ASElementMap.h"
+#import <UIKit/UIKit.h>
+#import <AsyncDisplayKit/ASCollectionElement.h>
+#import <AsyncDisplayKit/ASMultidimensionalArrayUtils.h>
+#import <AsyncDisplayKit/ASMutableElementMap.h>
+#import <AsyncDisplayKit/ASSection.h>
+#import <AsyncDisplayKit/NSIndexSet+ASHelpers.h>
+
+@interface ASElementMap ()
+
+@property (nonatomic, strong, readonly) NSArray<ASSection *> *sections;
+
+// Element -> IndexPath
+@property (nonatomic, strong, readonly) NSMapTable<ASCollectionElement *, NSIndexPath *> *elementToIndexPathMap;
+
+// The items, in a 2D array
+@property (nonatomic, strong, readonly) ASCollectionElementTwoDimensionalArray *sectionsOfItems;
+
+@property (nonatomic, strong, readonly) ASSupplementaryElementDictionary *supplementaryElements;
+
+@end
+
+@implementation ASElementMap
+
+- (instancetype)init
+{
+  return [self initWithSections:@[] items:@[] supplementaryElements:@{}];
+}
+
+- (instancetype)initWithSections:(NSArray<ASSection *> *)sections items:(ASCollectionElementTwoDimensionalArray *)items supplementaryElements:(ASSupplementaryElementDictionary *)supplementaryElements
+{
+  if (self = [super init]) {
+    _sections = [sections copy];
+    _sectionsOfItems = [[NSArray alloc] initWithArray:items copyItems:YES];
+    _supplementaryElements = [[NSDictionary alloc] initWithDictionary:supplementaryElements copyItems:YES];
+
+    // Setup our index path map
+    _elementToIndexPathMap = [NSMapTable mapTableWithKeyOptions:(NSMapTableStrongMemory | NSMapTableObjectPointerPersonality) valueOptions:NSMapTableCopyIn];
+    NSInteger s = 0;
+    for (NSArray *section in _sectionsOfItems) {
+      NSInteger i = 0;
+      for (ASCollectionElement *element in section) {
+        NSIndexPath *indexPath = [NSIndexPath indexPathForItem:i inSection:s];
+        [_elementToIndexPathMap setObject:indexPath forKey:element];
+        i++;
+      }
+      s++;
+    }
+    for (NSDictionary *supsOfKind in [_supplementaryElements objectEnumerator]) {
```

Comment:
Super super nit, feel free to ignore, I'd spell this out; it's the objective-c way :)